### DISCUSSION TITLE
[v3] allow multiple arguments on mutation function

### DISCUSSION THREAD

**Discussion by stefanprobst** [TanStack/query#1226] [Ideas]
**Describe the bug**

i would like to be able to pass multiple arguments (`TVariables[]`) to a `mutate` function, which currently fails with a type error. while it is possible to pack all arguments in one object, i think it makes sense to keep some things separate, e.g. for passing custom headers.

**To Reproduce**

1. create hook for mutation:

```ts
function useCreatePost() {
  return useMutation(
    (body: { title: string }, headers?: { [key: string]: string }) => {
      return fetch('https://jsonplaceholder.typicode.com/posts', {
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
          ...headers,
          'Content-type': 'application/json',
        },
      }).then((response) => response.json());
    }
  );
}
```

2. call hook with two arguments:

```ts
mutate({ title: 'A post' }, { Accept: 'application/json' })
```
which will result in type error: "Argument of type '{ Accept: string; }' is not assignable to parameter of type 'MutateOptions<any, unknown, { title: string; }, unknown>'."

**Expected behavior**
it would be cool if this would work in v3.

related: https://github.com/tannerlinsley/react-query/discussions/1085

**Comment by tannerlinsley** [comment]
We may consider this in v4. v3 is going to retain the current API style for now.

**Comment by boschni** [comment]
This is a request that we are getting more often and I do agree it is somewhat cumbersome to define separate interfaces for mutation functions.

It could be worthwhile to check how an alternative implementation might look like:
```js
function addComment(postId: string, comment: sting) {
  return api.post(`/posts/${postId}/comments`, { comment })
}

const { mutate } = useMutation(addComment)

mutate(1, 'comment')
```
The additional callback support on the `mutate` function would need to be dropped, but this might not be that bad as `mutateAsync` can also be used for those cases:

```js
function addComment(postId: string, comment: Comment) {
  return api.post(`/posts/${postId}/comments`)
}

const { mutateAsync } = useMutation(addComment)

try {
  await mutateAsync(1, 'comment')
  console.log('success')
} catch (error) {
  console.log('error')
} finally {
  console.log('settled')
}
```
What do you guys think @TkDodo @tannerlinsley ?

**Reply by stefanprobst** [reply]
would it be possible to support both, i.e. multiple args *and* callbacks with the approach suggested [here](https://github.com/tannerlinsley/react-query/discussions/1085#discussioncomment-84277)?

**Reply by boschni** [reply]
I'm not a big fan of using variadics for this, but it would also be incompatible with all TS versions prior to v4

**Reply by tannerlinsley** [reply]
I could be swayed to remove the `mutate(var, options)` overload. I agree that it *is possible* to achieve the same thing with normal async flow and probably should be that way. And TBH, if you really really prefer to have a secondary callback style, you could just use another useMutation if you really wanted to:

```js
// useAddComment.js
export function addComment(postId: string, comment: sting) {
  return api.post(`/posts/${postId}/comments`, { comment })
}

export const useAddComment = () => useMutation(addComment, {
  onSuccess: () => console.log('Hello 1')
})
```

```js
// Component.js
import { useAddComment } from './useAddComment'

function Component () {
  const { mutate } = useMutation(useAddComment().mutateAsync, {
    onSuccess: () => console.log('Hello 2')
  })

  mutate(1, 'comment')
}

```

**Reply by TkDodo** [reply]
hmm, does that mean the `mutate` function itself would no longer support options like the `onSuccess` callback?

```
function Component () {
  const { mutate } = useMutation(addComment)

  <button onClick={mutate('foo'), {
    onSuccess: () => console.log('Hello')
  })} />
```

**Reply by TkDodo** [reply]
The variadic tuple solution would be backwards compatible at runtime because the object syntax would just continue to work, as would the options. What would happen to projects that use TS 3.9? Would they get weak typings or a type error?

**Reply by boschni** [reply]
The mutate function would not have callbacks but `useMutation` would still have them. Older TS versions would error if you try to use the function, and the backwards compatibility is a bit flaky because if the last argument is an object, is it then a value or a config?

**Reply by tannerlinsley** [reply]
Yep. Obviously there are ways to get around that with duck-typing the config, but that's just a weird IMO. I think it would be safer to just stabilize the function signature.

**Reply by Omelyan** [reply]
> This is a request that we are getting more often

You get this request a lot because functions can have multiple arguments. it's convenient, it's type separation, in many cases wrapping all into one object is a bad idea.

**Reply by TkDodo** [reply]
This is a 2.5 year old thread. I haven't heard the request in 18 months 🤷

**Reply by Shaddix** [reply]
I can gently +1 to the initial request (so that there are 2 requests in 18 months!) :)
I'm autogenerating the mutations, and currently my generated `mutateFn` accepts the single parameter of BODY.

But when an API has query/URL/header parameters the users want to pass them to `mutate` function as well. So, multiple parameters would be of a great help here.

Here's a user [request thread](https://github.com/Shaddix/react-query-swagger/issues/28)

**Reply by githorse** [reply]
+1!

**Reply by felipedeboni** [reply]
@TkDodo, I also would like to be able to call as mutate(1, 'comment').

Apart from Typescript, it gets crystal clear to whoever is reading the code that 1 is the id that goes into url and the second parameter is the request body.

**Comment by TkDodo** [comment]
So, I am a bit biased here because I like the callbacks on the `mutate` function, but I'd like to outline why I like them:

1) co-location

We use it a lot to do things like close a dialog when the mutation is successful, for example:

```
const dialog = useDialog()
const [mutate] = useMutation(...)

<...lots-of-jsx-here...>

<button onClick={() => {
    mutate(variables, { onSuccess: dialog.close })
}}>Do it!</button>
```

In this case, I like that I can keep the onSuccess close to the onClick handler, so that I know what happens when I click that button. Yes, if I extract it to a function (e.g. `submitForm`)  it's all the same:

```
const dialog = useDialog()
const [mutate] = useMutation(...)
const submitForm = () => {
  mutate(variables, { onSuccess: dialog.close })
}

<...lots-of-jsx-here...>

<button onClick={submitForm}>Do it!</button>
```

so this is a minor point, but it would help me co-locate / inline some occasions.

2) closures

```
const [mutate] = useMutation(...)

<...lots-of-jsx-here...>

{
   items.map(item => (
       <button
           key={item.id}
           onClick={() => {
               mutate(item, { onSuccess: () => doSomethingWithId(item.id) })
           }}
         >
            Update {item.name}
       </button>
   )
}
```

I think the only solution to use the callback directly on `useMutation` here would be to extract the `<button>` to it's own component, pass the `item` and each instance would do it's own `useMutation`. Possible, but it seems like an unnecessary overhead just to be able to pass arguments to `useMutation` which I have in the closure

3) custom hooks

We often move mutations into custom hooks, to provide good namings and keep the actual references to our api out of the view:

```
// custom hook
const useUpdateItem = () => useMutation(({ id, ...payload }) => axios.post(`/item/${id}`, payload))

//in the view:

const dialog = useDialog()
const [updateItem] = useUpdateItem(...)

<button onClick={() => {
    updateItem(variables, { onSuccess: dialog.close })
}}>Do it!</button>

```

Without the callback, I would need to pass the `dialog` to the `useUpdateItem` hook just to be able to use the callback :/

4) same mutation, different callbacks

the obvious one, but has actually never happened to me:

```
const [updateItem] = useUpdateItem(...)

<button onClick={() => {
    updateItem(variables, { onSuccess: doSomething })
}}>Do it!</button>

<button onClick={() => {
    updateItem(variables, { onSuccess: doSomethingElse })
}}>Do something else!</button>
```

```
----------------
```

Can I do all of this with the async variant? Absolutely. I just need to wrap everything in try/catch, which some people just don't prefer, me included :)

So, for me, this are all bigger issues than the "we need multiple variables to useMutation" problem.

I think people request it for two reasons:

1) They have two variables, one being e.g. the id that goes to the url and one is the payload, and that should be separated.

This is also possible right now, like I hinted above:

```
const [mutate] = useMutation(({ id, ...payload }) => axios.post(`/item/${id}`, payload))

mutate({ id: 1, foo: "bar", baz: true })
// vs
mutate(1, { foo: "bar", baz: true })
```

I don't see this as a huge issue, because mostly, you will have `id` in a variable anyways, so it comes down to:

```
mutate({ id, foo: "bar", baz: true })
// vs
mutate(id, { foo: "bar", baz: true })
```

is that such a big deal, or am I missing something?

Imo, named parameters are even a good thing for readability, and I like that this is enforced by useMutation:

```
mutate(id, "foo", true, bar, 5)
```

yes, just don't do this if you don't like it. 😅 

2) keeping things separate, like in the above example (body and headers):

```
mutate({ title: 'A post' }, { Accept: 'application/json' })
```

an overhead, I agree:

```
mutate({ body: { title: 'A post' }, headers: { Accept: 'application/json' } })
```

and I also agree that typing individual parameters is usually easier in TypeScript than typing an object:

```
(body: { title: string }, headers?: Record<string, string>) => {
```

vs.

```
({ body, headers }: { body: { title: string }, headers?: Record<string, string> }) => {
```

Yuk, that is really ugly. Extracting the type obviously helps, and we are used to do that for React Components all day.

```
----------------
```

If callbacks-on-mutate and multiple-variables are mutually exclusive, I would prefer to keep the current notation.

Hope this wasn't too long and helps everyone to come to a conclusion :)

**Comment by tannerlinsley** [comment]
Sounds like we should keep the current syntax then and keep discussing.

**Comment by aliakbarazizi** [comment]
Another great solution that I found in [blog ](https://tkdodo.eu/blog/mastering-mutations-in-react-query#direct-updates) is

```
const useUpdateTitle = (id) => {

  const queryClient = useQueryClient()


  return useMutation(

    (newTitle) => axios.patch(`/posts/${id}`, { title: newTitle }),

    {

      // 💡 response of the mutation is passed to onSuccess

      onSuccess: (newPost) => {

        // ✅ update detail view directly

        queryClient.setQueryData(['posts', id], newPost)

      },

    }

  )

}
```

You can put id in the `useUpdateTitle` argument instead of mutation and its look like good.

But he only issue with this approach is you can't use same mutation in one component for different id, you kind of need to create component for each item and call this hook with id.

**Reply by Omelyan** [reply]
> Another great solution that I found

Aha. now try to use it with TypeScript...you'll be very unhappy))
In its simplest form it would look like: useMutation<void, void, DataSchema>(() => api);
and who put input schema on the third place... this is a mutation! it's input type should come first!
as @TkDodo  said: "The more your TypeScript code looks like JavaScript, the better".

so no good solution yet ;(

**Reply by TkDodo** [reply]
The important part is to type your MutationFunction - the rest can be inferred from there. Assuming that your input is the type `DataSchema` and the output is `Promise<DataSchemaResponse>`, I would do:

```
useMutation({
  mutationFn: (input: DataSchema): Promise<DataSchemaResponse> => ...,
  onSuccess (responseData) => {
    // responseData will be of type `DataSchemaResponse`
  }
})
```

this becomes obvious if you extract the `mutationFn` into a separate function:

```
function updateDataSchema(input: DataSchema): Promise<DataSchemaResponse> {
 ...
}

useMutation({
  mutationFn: updateDataSchema
})
```

using the notation that passes generics via `<>` is almost always wrong because it takes away type inference. I go into this topic (albeit about useQuery, but same principles apply) here: https://tkdodo.eu/blog/type-safe-react-query

**Reply by Omelyan** [reply]
> mutationFn: (input: DataSchema): Promise<DataSchemaResponse> => ...,

yeap, that's right. but the topic is about multiple arguments, so only this first example will work, and DataSchema will be mentioned twice: for mutationFn and updateDataSchema.

**Reply by TkDodo** [reply]
please show an example that doesn't work with typescript, I don't get it.

**Reply by palmmymumu** [reply]
To be clear, which mutation hook is better in your opinion?

```
export const useUpdateAccount = id => {
  const queryClient = useQueryClient()
  return useMutation(data => AccountService.update(id, data), {
    onSuccess: () => {
      queryClient.invalidateQueries(['accounts'])
      queryClient.invalidateQueries(['account', id])
    }
  })
}
```

And

```
export const useUpdateAccount = () => {
  const queryClient = useQueryClient()
  return useMutation(({ id, ...data }) => AccountService.update(id, data), {
    onSuccess: (_, params) => {
      queryClient.invalidateQueries(['accounts'])
      queryClient.invalidateQueries(['account', params.id])
    }
  })
}
```

The second one seems better for me as it can reuse with different id in same component.
But I'm not sure about other downsides.

**Reply by TkDodo** [reply]
> To be clear, which mutation hook is better in your opinion?

depends, and mostly doesn't matter. You'll have access to `id` wherever you call `useMutation` and likely wherever you call `mutate`.
